File: 백준/Silver/15666. N과 M （12）/N과 M （12）.cc
#include <iostream>
#include <algorithm>
using namespace std;
int n,m;
int arr[9];
int num[9];

void dfs(int h, int cnt) {
    if(cnt==m) {
        for(int i=0; i<m; i++) {
            cout<<arr[i]<<" ";
        }
        cout<<"\n";
        return ;
    }
    
    int xx=0;
    for(int i=h; i<n; i++) {
        if(xx!=num[i]) {
            arr[cnt]=num[i];
            xx=num[i];
            dfs(i,cnt+1);
        }
    }
}


int main() {
    cin>>n>>m;
    for(int i=0; i<n; i++) {
        cin>>num[i];
    }
    sort(num, num+n);
    dfs(0,0);
}

File: 백준/Silver/15666. N과 M （12）/N과 M （12）.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[] num;
    static int[] arr;
    //static int[] visited = new int[10001];
    static int n,m;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n];
        num = new int[m];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        dfs(0,0);


    }

    public static void dfs(int i, int d) {
        if(d == m) {
            for(int j = 0; j < m; j++) {
                System.out.print(num[j] + " ");
            }
            System.out.println();
            return;
        }

        int sameNum = 0;
        for(int j =i ; j < n; j++) {
            if(sameNum != arr[j]) {
                num[d] = arr[j];
                sameNum = arr[j];
                dfs(j, d+1);
            }
        }
    }
}
